internal/deploy/mitm/configuration.go

```diff
@@ -23,6 +23,117 @@ type Configuration struct {
+	// If unset, no path filtering is applied.
+	PathContains string
+	// The access token which must appear as a header.
+	// If unset, no access token filtering is applied.
+	AccessToken string
+	// The HTTP method which must be used for this to match.
+	// If unset, no HTTP method filtering is applied.
+	Method string
+}
+
+func (p FilterParams) FilterString() string {
+	// the filter is a python regexp
+	// "Regexes are Python-style" - https://docs.mitmproxy.org/stable/concepts-filters/
+	// re.escape() escapes very little:
+	// "Changed in version 3.7: Only characters that can have special meaning in a regular expression are escaped.
+	// As a result, '!', '"', '%', "'", ',', '/', ':', ';', '<', '=', '>', '@', and "`" are no longer escaped."
+	// https://docs.python.org/3/library/re.html#re.escape
+	//
+	// The majority of HTTP paths are just /foo/bar with % for path-encoding e.g @foo:bar=>%40foo%3Abar,
+	// so on balance we can probably just use the path directly.
+	var s strings.Builder
+	if p.PathContains != "" {
+		s.WriteString(" ~u .*" + p.PathContains + ".*")
+	}
+	if p.Method != "" {
+		s.WriteString(" ~m " + strings.ToUpper(p.Method))
+	}
+	if p.AccessToken != "" {
+		s.WriteString(" ~hq " + p.AccessToken)
+	}
+	return s.String()
+}
+
+// ConfigOpts specifies the desired configuration for mitmproxy
+type ConfigOpts struct {
+	// Which HTTP requests/responses should be caught for this configuration.
+	// Any HTTP request/response which does not meet this filter criteria will
+	// be passed through without invoking the callback function. If no Filter
+	// is provided, all HTTP requests/responses will be caught and the respective
+	// callback called for each.
+	Filter Filter
+	// The function to invoke when an HTTP request has met the filter
+	// criteria. This callback is invoked BEFORE the request reaches the server.
+	// If this callback function returns a response, the request will never
+	// reach the server and the response provided will be returned instead.
+	// If this callback function returns nil, or no request callback function
+	// is provided, requests will be passed to the server unaltered.
+	RequestCallback callback.Fn
+	// The function to invoke when an HTTP response has met the filter
+	// criteria. This callback is invoked AFTER the request has been processed
+	// by the server. If this callback function returns a response, the
+	// server response will never reach the client and the response provided
+	// will be returned instead. If this callback function returns nil, or no
+	// response callback function is provided, responses will be passed to the
+	// client unaltered.
+	ResponseCallback callback.Fn
+}
+
+// Apply the given configuration to mitmproxy, and call the `inner` function whilst
+// that configuration has been applied. mitmproxy will revert back to its default
+// configuration when `inner` returns.
+func (c *Configuration) Apply(opts ConfigOpts, inner func()) {
```

Comment:
In other languages, it would be idiomatic to call this sort of wrapper something like `WithConfigOpts`; for me, `Apply` makes me expect there to be lasting changes to the `Configuration`.

YMMV

Reply:
Good point, but then I think we need a better name than `ConfigOpts`.

Reply:
Renamed to `WithIntercept`.